chanfitness/revenue.go

```diff
@@ -0,0 +1,227 @@
+	// revenue record, return an empty one.
+	revenue, ok := record[pairChan]
+	if !ok {
+		return Revenue{}
+	}
+
+	return revenue
+}
+
+// setRevenue sets the revenue value for a target channel and its forwarding
+// pair. This function expects both maps to be initialized.
+func (r RevenueReport) setRevenue(targetChan, pairChan wire.OutPoint,
+	revenue Revenue) {
+
+	r.ChannelPairs[targetChan][pairChan] = revenue
+}
+
+// addUpstream gets the existing revenue record for the incoming channel with
+// the outgoing channel, adds the incomingChannel volume and fees and updates the
+// Revenue Report.
+func (r RevenueReport) addIncoming(incomingChannel,
+	outgoingChannel wire.OutPoint, amount, fees lnwire.MilliSatoshi) {
+	revenue := r.getRevenue(incomingChannel, outgoingChannel)
+
+	// Add the fees and revenue that have been earned to the existing revenue
+	// record.
+	revenue.AmountIncoming += amount
+	revenue.FeesIncoming += fees
+
+	log.Trace("Incoming channel: %v recording incoming volume: %v and fee: %v"+
+		" with outgoing channel: %v", incomingChannel, amount, fees, outgoingChannel)
+
+	// Set the new revenue record in the revenue report.
+	r.setRevenue(incomingChannel, outgoingChannel, revenue)
+}
+
+// addOutgoing gets the existing revenue record for the outgoing channel with
+// the incoming channel, adds the outgoing volume and fees and updates the
+// Revenue Report.
+func (r RevenueReport) addOutgoing(outgoingChannel,
+	incomingChannel wire.OutPoint, amount, fees lnwire.MilliSatoshi) {
+	revenue := r.getRevenue(outgoingChannel, incomingChannel)
+
+	// Add the fees and revenue that have been earned to the existing revenue
+	// record.
+	revenue.AmountOutgoing += amount
+	revenue.FeesOutgoing += fees
+
+	log.Trace("Outgoing channel: %v recording outgoing volume: %v and fee: %v"+
+		" with incoming channel: %v", outgoingChannel, amount, fees, incomingChannel)
+
+	// Set the new revenue record in the revenue report.
+	r.setRevenue(outgoingChannel, incomingChannel, revenue)
+}
+
+// revenueEvent provides the information captured by ForwardingEvents with
+// channel outpoints rather than short channel ids.
+type revenueEvent struct {
+	incomingChannel wire.OutPoint
+	outgoingChannel wire.OutPoint
+	incomingAmt     lnwire.MilliSatoshi
+	outgoingAmt     lnwire.MilliSatoshi
+}
+
+// revenueReport creates a revenue report for the set of events provided. It takes
+// an attribute incoming float which determines the fee spit between incoming and
+// outgoing channels.
+func revenueReport(events []revenueEvent, attributeIncoming float64) *RevenueReport {
+	record := &RevenueReport{
+		ChannelPairs: make(map[wire.OutPoint]map[wire.OutPoint]Revenue),
+	}
+
+	for _, event := range events {
+		// Calculate total fees earned for this event.
+		fee := event.incomingAmt - event.outgoingAmt
+
+		// Calculate fees earned by the incoming channel in this event.
+		incomingFees := lnwire.MilliSatoshi(float64(fee) * attributeIncoming)
```

Comment:
I'm wondering if we need `attributeIncoming` at all. When querying a specific channel, you get a pairwise breakdown of the total fees this channel was involved in.  There will be an overlap, but if I just want to get the total fees earned I'll get the report for all my channels and divide by 2.

Reply:
I think it's more intuitive to have it already divided by 2? It doesn't make sense to double report by default imo.

Reply:
I am not sure about the `/2` either. It is kind of implicit that that is happening. The factor is even configurable, but would that be something that users need?